### DISCUSSION TITLE
Add option to encrypt fields (using a secret key)

### DISCUSSION THREAD

**Discussion by binaryfire** [directus/directus#8017] [Feature Requests]
Hi guys

Just opening an issue for this as per our Twitter conversation: https://twitter.com/binaryfire/status/1435440836162981893

We're currently using a combination of Hashicorp Vault and Directus to store data (Directus for non-sensitive fields and Vault for sensitive fields).

Could you add the ability to encrypt fields in Directus? This would allow us to store everything in Directus. Field-level encryption will protect sensitive data in the event of a database breach and more importantly, will allow us to comply with certain industry regulations that require sensitive data to be encrypted at rest.  

Here are some thoughts on how this might be implemented:

- Allow any field type to be encrypted (rather than providing a separate "encrypted string" interface). We'd like to be able to encrypt data in any format (strings, code blocks, WYSIWYG HTML etc).

- Perhaps the existing SECRET environment variable could be used as the encryption key? There could be a separate "ENCRYPTION_KEY" variable too

**Comment by rijkvanzanten** [comment]
@binaryfire What would you like/expect Directus to use to encrypt/decrypt the fields? "Regular" aes-256-ctr using Node's built-in crypto, or something else? 🤔

**Comment by binaryfire** [comment]
@rijkvanzanten I think the node crypto function uses OpenSSH's crypto functions. You should be able to use a variety of ciphers (aes-256-cbc, aes-256-ctr etc). So it probably makes sense to use node's built in functionality rather than adding another dependency.

I'm not sure which algorithm is best though. Here's a good Stack Overflow discussion: https://stackoverflow.com/questions/1220751/how-to-choose-an-aes-encryption-mode-cbc-ecb-ctr-ocb-cfb

What does node use by default?

**Reply by rijkvanzanten** [reply]
I don't think Node has a default cipher, as you explicitly pass the cipher to use to node's crypto function. I think the best way forwards is to expose that cipher through an environment variable, so the use can choose which one to use (and we can default to whichever one we deem most appropriate)

**Comment by biapar** [comment]
This is a GDPR rule. Is’s cypher key by user? You can store it in other database or in a global collection, but how to protect it?

**Reply by binaryfire** [reply]
I think the cipher key should be stored as an ENV variable (in the .env file). That's much more secure than storing it in the database.

The key wouldn't be provided by the user. It's store at the app level. The user is already authenticating with a password + 2FA (or token).

The goal with encryption isn't to strengthen authentication. It's to encrypt the data at rest.

**Reply by biapar** [reply]
Where to create the code to encrypt/decrypt some collection fields? In event before save and after read?

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
